### PR TITLE
13.0 mig account cutoff prepaid

### PR DESCRIPTION
**Pull request by jimhoefnagels**
depends #120 #125

### CODE REVIEW

**Review by TDu — APPROVED** [review]
LG

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+                    "Don't forget to Re-Generate Lines after entering or "
+                    "leaving forecast mode."
+                ),
+            }
+        }
+
+    def _prepare_prepaid_lines(self, aml, mapping):
+        self.ensure_one()
+        start_date_dt = fields.Date.from_string(aml.start_date)
+        end_date_dt = fields.Date.from_string(aml.end_date)
```
Comment (i-vyshnevska) [inline account_cutoff_prepaid/models/account_cutoff.py]: no  need to convert aml.start_date already data type

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        # Here, we compute the amount of the cutoff
+        # That's the important part !
+        total_days = (end_date_dt - start_date_dt).days + 1
+        if self.forecast:
+            out_days = 0
+            forecast_start_date_dt = fields.Date.from_string(self.start_date)
+            forecast_end_date_dt = fields.Date.from_string(self.end_date)
```
Comment (i-vyshnevska) [inline account_cutoff_prepaid/models/account_cutoff.py]: same

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                out_days += (forecast_start_date_dt - start_date_dt).days
+            prepaid_days = total_days - out_days
+        else:
+            cutoff_date_str = self.cutoff_date
+            cutoff_date_dt = fields.Date.from_string(cutoff_date_str)
+            if start_date_dt > cutoff_date_dt:
+                prepaid_days = total_days
+            else:
+                prepaid_days = (end_date_dt - cutoff_date_dt).days
+        assert total_days > 0, "Should never happen. Total days should always be > 0"
```
Comment (i-vyshnevska) [inline account_cutoff_prepaid/models/account_cutoff.py]: ValidationError?

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        company = self.env.user.company_id
+        if cutoff_type == "prepaid_revenue":
+            account_id = company.default_prepaid_revenue_account_id.id or False
+        elif cutoff_type == "prepaid_expense":
+            account_id = company.default_prepaid_expense_account_id.id or False
+        return account_id
+
+    @api.model
+    def create(self, vals):
+        res = super(AccountCutoff, self).create(vals)
```
Comment (i-vyshnevska) [inline account_cutoff_prepaid/models/account_cutoff.py]:
```suggestion
        res = super().create(vals)
```

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+
+class AccountCutoffLine(models.Model):
+    _inherit = "account.cutoff.line"
+
+    move_line_id = fields.Many2one(
+        "account.move.line", string="Account Move Line", readonly=True
+    )
+    move_date = fields.Date(
+        related="move_line_id.date", string="Account Move Date", readonly=True
```
Comment (i-vyshnevska) [inline account_cutoff_prepaid/models/account_cutoff.py:185]:
```suggestion
        related="move_line_id.date", string="Account Move Date"
```

**Comment by i-vyshnevska** [comment]
@jimhoefnagels thanks for changes, can you apply pre-commit on PR

File: account_cutoff_prepaid/models/account_cutoff.py [hunk 6]
```diff
@@ -97,7 +95,8 @@ def _prepare_prepaid_lines(self, aml, mapping):
                 prepaid_days = total_days
             else:
                 prepaid_days = (end_date_dt - cutoff_date_dt).days
-        assert total_days > 0, "Should never happen. Total days should always be > 0"
+        if total_days > 0:
+            raise ValidationError(_("Total days should always be > 0"))
```
Comment (alexis-via) [inline account_cutoff_prepaid/models/account_cutoff.py:99]: Arf, that's a big mistake. It seems you didn't test your v13 migration in real life !
I'm preparing a PR to fix all this and get the improvements from v12 too.

**Pull request merged**